platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/DefaultDependencyManagementServices.java

```diff
@@ -171,18 +175,69 @@ public DefaultDependencyManagementServices(ServiceRegistry parent) {
     }
 
     @Override
-    public DependencyResolutionServices create(FileResolver resolver, FileCollectionFactory fileCollectionFactory, DependencyMetaDataProvider dependencyMetaDataProvider, ProjectFinder projectFinder, DomainObjectContext domainObjectContext) {
+    public DependencyResolutionServices newDetachedResolver(
+        FileResolver resolver,
+        FileCollectionFactory fileCollectionFactory,
+        DomainObjectContext owner
+    ) {
+        DependencyResolutionServices services = newDetachedResolver(
+            resolver,
+            fileCollectionFactory,
+            owner,
+            new AnonymousModule()
+        );
+
+        // We restrict this so that detached resolvers only represent adhoc root components that do not expose variants.
+        services.getConfigurationContainer().configureEach(configuration -> {
+            if (configuration.isCanBeConsumed()) {
+                throw new InvalidUserCodeException("Cannot create consumable configurations in detached resolvers");
+            }
```

Comment:
❓ Will this catch every case of consumable configurations?  Can't a script create a non-consumable configuration and then register its own `configureEach` block (which will be run after this one) to make it consumable?

Reply:
It is technically possible to bypass this, but `newDetachedResolver` is only used by internal code so this is more of a safeguard from ourselves. This should protect from most accidental creations of consumable configurations except those that really try to abuse this.